accumulator_updater/programs/accumulator_updater/src/lib.rs

```diff
@@ -0,0 +1,269 @@
+        require_keys_neq!(allowed_program, Pubkey::default());
+        require!(
+            !whitelist.allowed_programs.contains(&allowed_program),
+            AccumulatorUpdaterError::DuplicateAllowedProgram
+        );
+        whitelist.allowed_programs.push(allowed_program);
+        Ok(())
+    }
+
+    /// Add new account(s) to be included in the accumulator
+    pub fn add_account<'info>(
+        ctx: Context<'_, '_, '_, 'info, AddAccount<'info>>,
+        base_account: Pubkey,
+        data: Vec<Vec<u8>>,
+        account_type: u32,
+        account_schemas: Vec<u8>,
+    ) -> Result<()> {
+        let cpi_caller = ctx.accounts.whitelist_verifier.is_allowed()?;
+        require_eq!(data.len(), account_schemas.len());
+        let accts = ctx.remaining_accounts;
+        require_eq!(accts.len(), data.len());
+        let mut zip = data.into_iter().zip(account_schemas.into_iter());
+
+        for ai in accts {
+            let (account_data, account_schema) = zip.next().unwrap();
+            let seeds = accumulator_acc_seeds!(cpi_caller, base_account, account_schema);
+            let (pda, bump) = Pubkey::find_program_address(seeds, &crate::ID);
+            require_keys_eq!(ai.key(), pda);
+
+            let accumulator_size = 8 + AccumulatorInput::get_initial_size(&account_data);
+            let accumulator_account = AccumulatorInput {
+                header: AccumulatorHeader {
+                    version: 1, // from cpi caller?
+                    account_type,
+                    account_schema,
+                },
+                data:   account_data,
+            };
+            create_and_initialize_accumulator_account_pda(
+                ai,
+                accumulator_account,
+                accumulator_size,
+                &ctx.accounts.payer,
+                &[accumulator_acc_seeds_with_bump!(
+                    cpi_caller,
+                    base_account,
+                    account_schema,
+                    bump
+                )],
+                &ctx.accounts.system_program,
+            )?;
+        }
+
+        Ok(())
+    }
+}
+
+fn create_and_initialize_accumulator_account_pda<'a>(
+    accumulator_input_ai: &AccountInfo<'a>,
+    accumulator_input: AccumulatorInput,
+    accumulator_input_size: usize,
+    payer: &AccountInfo<'a>,
+    seeds: &[&[&[u8]]],
+    system_program: &AccountInfo<'a>,
+) -> Result<()> {
+    let rent = Rent::get()?;
+    let lamports = rent.minimum_balance(accumulator_input_size);
+    let create_pda_ix = &system_instruction::create_account(
+        payer.key,
+        accumulator_input_ai.key,
+        lamports,
+        accumulator_input_size.try_into().unwrap(),
+        &crate::ID,
+    );
+
+    invoke_signed(
+        create_pda_ix,
+        &[
+            payer.clone(),
+            accumulator_input_ai.clone(),
+            system_program.clone(),
+        ],
+        seeds,
+    )?;
+
+
+    AccountSerialize::try_serialize(
```

Comment:
this uses the anchor framework for serde which uses BorshSerialization by default and the first 8 bytes will be an account discriminator/identifier for security purposes. This was just to get an initial implementation working since the deserialization logic is automatically handled/generated for the typescript client, but I will add a TODO to handle this for this.

Reply:
We usually manually hand-roll the data for cross chain usage as different frameworks are entirely different. I suggest doing the same here. For example on EVM everything is big-endian which is different than other enviornments.